chainer/distribution.py

```diff
@@ -0,0 +1,420 @@
+            x(:class:`~chainer.Variable` or :class:`numpy.ndarray` or \
+            :class:`cupy.ndarray`): Data points in the domain of the
+            distribution
+
+        Returns:
+            ~chainer.Variable: Logarithm of survival function value evaluated
+            at `x`.
+
+        """
+        raise NotImplementedError
+
+    @property
+    def mean(self):
+        """Returns the mean of the distribution.
+
+        Returns:
+            ~chainer.Variable: The mean of the distribution.
+
+        """
+        raise NotImplementedError
+
+    @property
+    def mode(self):
+        """Returns the mode of the distribution.
+
+        Returns:
+            ~chainer.Variable: The mode of the distribution.
+
+        """
+        raise NotImplementedError
+
+    def perplexity(self, x):
+        """Evaluates the perplexity function at a given points.
+
+        Args:
+            x(:class:`~chainer.Variable` or :class:`numpy.ndarray` or \
+            :class:`cupy.ndarray`): Data points in the domain of the
+            distribution
+
+        Returns:
+            ~chainer.Variable: Perplexity function value evaluated at `x`.
+
+        """
+        raise NotImplementedError
+
+    def prob(self, x):
+        """Evaluates probability at a given points.
+
+        Args:
+            x(:class:`~chainer.Variable` or :class:`numpy.ndarray` or \
+            :class:`cupy.ndarray`): Data points in the domain of the
+            distribution
+
+        Returns:
+            ~chainer.Variable: Probability evaluated at `x`.
+
+        """
+        raise NotImplementedError
+
+    def sample(self, sample_shape=()):
+        """Samples random points from the distribution.
+
+        This function call `sample_n` and reshape a result of `sample_n` to
+        `(sample_shape, batch_shape, event_shape)`. On implementing sampling
+        codes in an inherited ditribution class, it is deprecated to override
+        this function. Instead to do this, it is preferable to override
+        `sample_n`.
+
+        Args:
+            sample_shape(:class:`tuple` of :class:`int`): Sampling shape.
+
+        Returns:
+            ~chainer.Variable: Sampled random points.
+        """
+        final_shape = self.batch_shape + self.event_shape
+        if sample_shape == ():
+            n = 1
+        elif isinstance(sample_shape, int):
+            n = sample_shape
+            final_shape = (n,) + final_shape
+        else:
+            n = 1
+            for shape_ in sample_shape:
+                n *= shape_
+            final_shape = sample_shape + final_shape
+        samples = self.sample_n(n)
+        return samples.reshape(final_shape)
+
+    def sample_n(self, n):
+        """Samples n random points from the distribution.
+
+        This function returns sampled points whose shape is
+        `(n, batch_shape, event_shape)`. On implementing sampling codes in an
+        inherited ditribution class, it is preferable to override this method.
```

Comment:
Consider rewording, such as: When implementing sampling code in a subclass, it is recommended to override this method.